<a href="https://colab.research.google.com/github/karolinavg/Evidencias-de-las-Actividades/blob/main/clase7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Regression
25/nov/2022

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
data='https://raw.githubusercontent.com/tec03/Datasets/main/datasets/reg_data.csv'

In [ ]:
df=pd.read_csv(data)
df

Gender  Salary   Age       Place  Weight Company   Education  JoinYear
0   Female  1300.0  35.0     Chicago    80.0     BMW    Bachelor      2008
1   Female  1250.0  33.0     Chicago    82.5    Ford  HighSchool      2010
2     Male  2200.0  34.0    New York   100.8     BMW    Bachelor      2004
3     Male  2100.0  42.0    New York    90.0     BMW      Master      2000
4   Female  1500.0  29.0     Chicago    67.0    Ford      Master      2010
5   Female  1700.0  42.0  Washington    60.0    Ford      Master      2000
6     Male  3000.0  50.0  Washington    77.0    Ford  HighSchool      2010
7     Male  3000.0  55.0  Washington    77.0    Ford    Bachelor      2010
8   Female  2800.0  31.0    New York    87.0    Ford    Bachelor      2004
9     Male  2900.0  46.0    New York    70.0      GM      Master      2008
10  Female  2780.0  36.0  Washington    57.0     BMW  HighSchool      2009
11    Male  2550.0  48.0    New York    64.0      GM      Master      2004
12  Female     NaN  31.0    New York    87.0    Ford    Bachelor      2009
13    Male  2900.0   NaN    New York    70.0      GM      Master      2010
14  Female  2780.0  36.0         NaN    57.0     BMW  HighSchool      2010
15    Male  2550.0  48.0    New York     NaN      GM      Master      2000
16    Male  2550.0  48.0    New York    64.0     NaN      Master      2000
17    Male  2550.0  48.0    New York    64.0      GM         NaN      2000
18     NaN  2550.0  48.0    New York    64.0      GM      Master      2009

## Description of the variables
*   Gender: gives information about the gender of each employees. It is a categorical variable.



In [ ]:
ndf=df.copy()
ndf.dropna(inplace=True)

In [ ]:
ndf

Gender  Salary   Age       Place  Weight Company   Education  JoinYear
0   Female  1300.0  35.0     Chicago    80.0     BMW    Bachelor      2008
1   Female  1250.0  33.0     Chicago    82.5    Ford  HighSchool      2010
2     Male  2200.0  34.0    New York   100.8     BMW    Bachelor      2004
3     Male  2100.0  42.0    New York    90.0     BMW      Master      2000
4   Female  1500.0  29.0     Chicago    67.0    Ford      Master      2010
5   Female  1700.0  42.0  Washington    60.0    Ford      Master      2000
6     Male  3000.0  50.0  Washington    77.0    Ford  HighSchool      2010
7     Male  3000.0  55.0  Washington    77.0    Ford    Bachelor      2010
8   Female  2800.0  31.0    New York    87.0    Ford    Bachelor      2004
9     Male  2900.0  46.0    New York    70.0      GM      Master      2008
10  Female  2780.0  36.0  Washington    57.0     BMW  HighSchool      2009
11    Male  2550.0  48.0    New York    64.0      GM      Master      2004

### Salary of a person with age 33?
Salary=b0+b1*Age

In [ ]:
import statsmodels.api as sm

In [ ]:
favs=['Salary','Age']
ndf=df[favs]
ndf.dropna(inplace=True)
ndf

/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


Salary   Age
0   1300.0  35.0
1   1250.0  33.0
2   2200.0  34.0
3   2100.0  42.0
4   1500.0  29.0
5   1700.0  42.0
6   3000.0  50.0
7   3000.0  55.0
8   2800.0  31.0
9   2900.0  46.0
10  2780.0  36.0
11  2550.0  48.0
14  2780.0  36.0
15  2550.0  48.0
16  2550.0  48.0
17  2550.0  48.0
18  2550.0  48.0

In [ ]:
import statsmodels.api as sample_data
from sklearn import linear_model as LM

In [ ]:
#Salary=b0+b1*Age
x=pd.DataFrame(ndf.Age)
y=pd.DataFrame(ndf.Salary)

In [ ]:
type(x)

pandas.core.frame.DataFrame

In [ ]:
type(y)

pandas.core.frame.DataFrame

In [ ]:
lm=LM.LinearRegression()
model_lm=lm.fit(x,y)
model_lm

LinearRegression()

In [ ]:
print('Model intercept is: ',model_lm.intercept_)
print('Model coefficient is: ',model_lm.coef_)
model_lm.score(x,y)

Model intercept is:  [606.34509569]
Model coefficient is:  [[41.96351675]]


0.31711161647425

### What is the impact of Age in salary?
salary=b0 + b1*age



In [ ]:
model1A=sm.OLS.from_formula('Salary~Age',data=df)
resultA=model1A.fit()
resultA.summary()

/usr/local/lib/python3.7/dist-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=17
  "anyway, n=%i" % int(n))


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.317
Model:                            OLS   Adj. R-squared:                  0.272
Method:                 Least Squares   F-statistic:                     6.966
Date:                Tue, 29 Nov 2022   Prob (F-statistic):             0.0186
Time:                        18:09:26   Log-Likelihood:                -128.66
No. Observations:                  17   AIC:                             261.3
Df Residuals:                      15   BIC:                             263.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    606.3451    674.058      0.900      0.383    -830.376    2043.066
Age           41.9635     15.900      2.639      0.019       8.074      75.853
==============================================================================
Omnibus:                        0.097   Durbin-Watson:                   1.273
Prob(Omnibus):                  0.953   Jarque-Bera (JB):                0.315
Skew:                           0.084   Prob(JB):                        0.854
Kurtosis:                       2.354   Cond. No.                         236.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Salary =b0 + b1*age
### Salary =606.3451 + 41.9635*age

In [ ]:
print('Salary of a person with age 33 is',606.3451 + 41.9635*33)

Salary of a person with age 33 is 1991.1406000000002


In [ ]:
ndf

Salary   Age
0   1300.0  35.0
1   1250.0  33.0
2   2200.0  34.0
3   2100.0  42.0
4   1500.0  29.0
5   1700.0  42.0
6   3000.0  50.0
7   3000.0  55.0
8   2800.0  31.0
9   2900.0  46.0
10  2780.0  36.0
11  2550.0  48.0
14  2780.0  36.0
15  2550.0  48.0
16  2550.0  48.0
17  2550.0  48.0
18  2550.0  48.0

In [ ]:
resultA.resid

0    -775.068182
1    -741.141148
2     166.895335
3    -268.812799
4    -323.287081
5    -668.812799
6     295.479067
7      85.661483
8     892.785885
9     363.333134
10    662.968301
11    -70.593900
14    662.968301
15    -70.593900
16    -70.593900
17    -70.593900
18    -70.593900
dtype: float64

### How the age and gender impact de salary?
Salary=b0 + b1*age, b2 * gender

In [ ]:
model1B=sm.OLS.from_formula('Salary~ Age + Gender',data=df)
resultB=model1B.fit()
resultB.summary()

/usr/local/lib/python3.7/dist-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  "anyway, n=%i" % int(n))


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.323
Model:                            OLS   Adj. R-squared:                  0.219
Method:                 Least Squares   F-statistic:                     3.108
Date:                Tue, 29 Nov 2022   Prob (F-statistic):             0.0789
Time:                        18:17:17   Log-Likelihood:                -121.44
No. Observations:                  16   AIC:                             248.9
Df Residuals:                      13   BIC:                             251.2
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        874.7433    968.174      0.903      0.383   -1216.869    2966.355
Gender[T.Male]   188.7701    423.620      0.446      0.663    -726.406    1103.946
Age               33.0033     27.396      1.205      0.250     -26.183      92.189
==============================================================================
Omnibus:                        0.559   Durbin-Watson:                   1.244
Prob(Omnibus):                  0.756   Jarque-Bera (JB):                0.623
Skew:                           0.295   Prob(JB):                        0.732
Kurtosis:                       2.234   Cond. No.                         320.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Salary=b0 + b1*age, b2 * gender

Salary = 874.7433 + 188.7701*gender + 33.0033*age 

In [ ]:
print('Salary of a female with age 33 is',874.7433 + 188.7701*0 + 33.0033*33)

Salary of a female with age 33 is 1963.8522000000003


###Salary of a female with age 33 working in BMW is

In [ ]:
model1C=sm.OLS.from_formula('Salary~ Age + Gender + Company',data=df)
resultC=model1C.fit()
resultC.summary()

/usr/local/lib/python3.7/dist-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  "anyway, n=%i" % int(n))


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.329
Model:                            OLS   Adj. R-squared:                  0.060
Method:                 Least Squares   F-statistic:                     1.225
Date:                Tue, 29 Nov 2022   Prob (F-statistic):              0.360
Time:                        18:17:12   Log-Likelihood:                -114.21
No. Observations:                  15   AIC:                             238.4
Df Residuals:                      10   BIC:                             242.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept         783.6221   1129.351      0.694      0.504   -1732.729    3299.974
Gender[T.Male]    143.1703    528.628      0.271      0.792   -1034.686    1321.027
Company[T.Ford]  -143.3508    390.332     -0.367      0.721   -1013.065     726.364
Company[T.GM]     -94.6944    484.605     -0.195      0.849   -1174.461     985.072
Age                38.0085     33.731      1.127      0.286     -37.149     113.166
==============================================================================
Omnibus:                        0.350   Durbin-Watson:                   1.205
Prob(Omnibus):                  0.839   Jarque-Bera (JB):                0.476
Skew:                           0.255   Prob(JB):                        0.788
Kurtosis:                       2.293   Cond. No.                         318.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**SALARY = b0 + b1*gender + b2*age + b3*company(ford) + b4*company(GM)**
* SALARY = 783.6221 + 143.1703*gender + 38.0085*age - 143.3508*company(ford) - 94.6944*company(GM)

In [ ]:
print('Salary of a female with age 33 working in BMW is', 783.6221 + 143.1703*0 + 38.0085*33 - 143.3508*0 - 94.6944*0)

Salary of a female with age 33 working in BMW is 2037.9026


In [ ]:
print('Salary of a female with age 33 working in Ford is', 783.6221 + 143.1703*0 + 38.0085*33 - 143.3508*1 - 94.6944*0)

Salary of a female with age 33 working in BMW is 1894.5518


### Salary of a female with age 33 working in Ford joined 2009

In [ ]:
model1D=sm.OLS.from_formula('Salary~ Age + Gender + Company + JoinYear',data=df)
resultD=model1D.fit()
resultD.summary()

/usr/local/lib/python3.7/dist-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  "anyway, n=%i" % int(n))


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.374
Model:                            OLS   Adj. R-squared:                  0.026
Method:                 Least Squares   F-statistic:                     1.074
Date:                Tue, 29 Nov 2022   Prob (F-statistic):              0.435
Time:                        18:35:52   Log-Likelihood:                -113.70
No. Observations:                  15   AIC:                             239.4
Df Residuals:                       9   BIC:                             243.6
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept       -6.864e+04   8.65e+04     -0.793      0.448   -2.64e+05    1.27e+05
Gender[T.Male]    212.7294    545.242      0.390      0.705   -1020.695    1446.154
Company[T.Ford]  -169.4085    398.803     -0.425      0.681   -1071.563     732.746
Company[T.GM]       1.8537    507.937      0.004      0.997   -1147.180    1150.887
Age                35.4906     34.492      1.029      0.330     -42.535     113.516
JoinYear           34.6373     43.172      0.802      0.443     -63.026     132.300
==============================================================================
Omnibus:                        1.431   Durbin-Watson:                   1.281
Prob(Omnibus):                  0.489   Jarque-Bera (JB):                0.355
Skew:                           0.355   Prob(JB):                        0.837
Kurtosis:                       3.253   Cond. No.                     1.10e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.1e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**SALARY = b0 + b1*gender + b2*age + b3*company(ford) + b4*company(GM) * B5*JoinYear**
* SALARY = -6.864e+04	 + 212.7294	*gender[male] + 35.4906*age - 169.4085*company[Ford] + 	1.8537*company[GM] + 34.6373*JoinYear **texto en negrita**

In [ ]:
print('The salary of a 33 year old female working in FORD since 2009 is',-6.864e+04	 + 212.7294	*0 + 35.4906*33 - 169.4085*1 + 	1.8537*0 + 34.6373*2010)

The salary of a 33 year old female working in FORD since 2009 is 1982.7543000000005


In [ ]:
df.dtypes

Gender        object
Salary       float64
Age          float64
Place         object
Weight       float64
Company       object
Education     object
JoinYear       int64
dtype: object

In [ ]:
df.JoinYear.unique()

array([2008, 2010, 2004, 2000, 2009])

In [ ]:
# LA VARIABLE JOIN YEAR ES CATEGORICA PERO PYTHON NO LO SABE, ENTONCES TENEMOS QUE CAMBIARLA 
df.JoinYear=df.JoinYear.astype('str')


In [ ]:
model1E=sm.OLS.from_formula('Salary~ Age + Gender + Company + JoinYear',data=df)
resultE=model1E.fit()
resultE.summary()

/usr/local/lib/python3.7/dist-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  "anyway, n=%i" % int(n))


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.571
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.9985
Date:                Tue, 29 Nov 2022   Prob (F-statistic):              0.516
Time:                        18:49:02   Log-Likelihood:                -110.86
No. Observations:                  15   AIC:                             239.7
Df Residuals:                       6   BIC:                             246.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         -110.7400   1431.430     -0.077      0.941   -3613.324    3391.844
Gender[T.Male]      62.5265    660.624      0.095      0.928   -1553.963    1679.016
Company[T.Ford]   -149.2487    508.569     -0.293      0.779   -1393.672    1095.174
Company[T.GM]       76.3867    537.625      0.142      0.892   -1239.135    1391.909
JoinYear[T.2004]   694.8991    539.822      1.287      0.245    -625.999    2015.797
JoinYear[T.2008]    82.1158    564.640      0.145      0.889   -1299.510    1463.741
JoinYear[T.2009]  1060.3688    802.089      1.322      0.234    -902.273    3023.011
JoinYear[T.2010]   446.8765    494.418      0.904      0.401    -762.921    1656.674
Age                 50.8436     41.630      1.221      0.268     -51.022     152.709
==============================================================================
Omnibus:                        0.333   Durbin-Watson:                   1.595
Prob(Omnibus):                  0.847   Jarque-Bera (JB):                0.329
Skew:                           0.281   Prob(JB):                        0.848
Kurtosis:                       2.540   Cond. No.                         403.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**SALARY = b0 + b1*gender + b2*age + b3*company(ford) + b4*company(GM) * b5*JoinYear[2004] + b6*JoinYear[2008] + b7*JoinYear[2009] + b8*JoinYear[2010]**
* SALARY = 	-110.7400 + 62.5265*gender[male] + 50.8436*age - 149.2487*company[Ford] +	76.3867*company[GM] + 34.6373*JoinYear[2004] **texto en negrita**